components/cam/src/physics/cam/micro_p3.F90

```diff
@@ -920,226 +919,46 @@ SUBROUTINE p3_main(qc,nc,qr,nr,th,qv,dt,qitot,qirim,nitot,birim,   &
 
           !................................................................
           ! deposition/condensation-freezing nucleation
-          ! allow ice nucleation if < -15 C and > 5% ice supersaturation
-          ! use CELL-AVERAGE values, freezing of vapor
-
-          if ( t(i,k).lt.icenuct .and. supi(i,k).ge.0.05_rtype) then
-            if(.not. log_predictNc) then 
-               ! dum = exp(-0.639+0.1296*100.*supi(i,k))*1000.*inv_rho(i,k)  !Meyers et al. (1992)
-               dum = 0.005_rtype*exp(0.304_rtype*(zerodegc-t(i,k)))*1000._rtype*inv_rho(i,k)   !Cooper (1986)
-               dum = min(dum,100.e3_rtype*inv_rho(i,k))
-               N_nuc = max(0._rtype,(dum-nitot(i,k))*odt)
-               if (N_nuc.ge.1.e-20_rtype) then
-                  Q_nuc = max(0._rtype,(dum-nitot(i,k))*mi0*odt)
-                  qinuc = Q_nuc
-                  ninuc = N_nuc
-               endif
-            else 
-            ! Ice nucleation predicted by aerosol scheme 
-               ninuc = max(0._rtype, (naai(i,k) - nitot(i,k))*odt)
-               qinuc = ninuc * mi0
-            endif 
-         endif 
+          call ice_nucleation(t(i,k),inv_rho(i,k),& 
+            nitot(i,k),naai(i,k),supi(i,k),odt,log_predictNc,& 
+            qinuc, ninuc)
+
           !.................................................................
           ! droplet activation
-
-          if (log_predictNc) then
-            ! for predicted Nc, use activation predicted by aerosol scheme
-            ! note that this is also applied at the first time step
-            if (sup(i,k).gt.1.e-6) then
-               ncnuc = npccn(i,k)
-               if (it.eq.1) then
-                  qcnuc = 0._rtype
-               else
-                  !TODO Limit qcnuc so that conditions never become sub-saturated 
-                  qcnuc = ncnuc*cons7
-               endif
-            endif
-         else if (sup(i,k).gt.1.e-6.and.it.gt.1) then
-           ! for specified Nc, make sure droplets are present if conditions are supersaturated
-           ! this is not applied at the first time step, since saturation adjustment is applied at the first step
-            dum   = nccnst*inv_rho(i,k)*cons7-qc(i,k)
-            dum   = max(0._rtype,dum)
-            dumqvs = qv_sat(t(i,k),pres(i,k),0)
-            dqsdt = xxlv(i,k)*dumqvs/(rv*t(i,k)*t(i,k))
-            ab    = 1._rtype + dqsdt*xxlv(i,k)*inv_cp
-            dum   = min(dum,(qv(i,k)-dumqvs)/ab)  ! limit overdepletion of supersaturation
-            qcnuc = dum*odt
-         endif
+          call droplet_activation(t(i,k),pres(i,k),qv(i,k),qc(i,k),inv_rho(i,k),& 
```

Comment:
More of a sanity check, are we sure that droplet activation should be computed using cell-averaged values?

Reply:
@AaronDonahue I think this is ok. When we remove the first step treatment of droplet activation, I think the droplet_activation subroutine will no longer depend on qc.

Reply:
Actually now that I look more closely at the `if` statement in the routine, we only care about `qc` when `log_predictNc` is turned off.  Which is not our baseline case anyways.